include/ck/tensor_operation/gpu/block/blockwise_gemm_xdlops.hpp

```diff
@@ -255,58 +259,82 @@ struct BlockwiseGemmXdlops_k0mk1_k0nk1_m0n0m1n1m2m3m4n2_v1
         auto b_thread_buf = make_static_buffer<AddressSpaceEnum::Vgpr, FloatAB>(
             b_thread_desc_.GetElementSpaceSize());
 
-        static_for<0, MRepeat, 1>{}([&](auto m0) {
-            // read A
-            a_thread_copy_.Run(a_block_desc_m0_m1_m2_k,
-                               make_tuple(m0, I0, I0, I0),
-                               a_block_buf,
-                               a_thread_desc_,
-                               make_tuple(I0, I0, I0, I0),
-                               a_thread_buf);
-
+        static_for<0, KPerBlock, KPerInnerLoop>{}([&](auto k) {
+            static_for<0, MRepeat, 1>{}([&](auto m0) {
+                // read A
+                a_thread_copy_.Run(a_block_desc_m0_m1_m2_k,
+                                   make_tuple(m0, I0, I0, k),
+                                   a_block_buf,
+                                   a_thread_desc_,
+                                   make_tuple(m0, I0, I0, I0),
+                                   a_thread_buf);
+            });
             static_for<0, NRepeat, 1>{}([&](auto n0) {
                 // read B
                 b_thread_copy_.Run(b_block_desc_n0_n1_n2_k,
-                                   make_tuple(n0, I0, I0, I0),
+                                   make_tuple(n0, I0, I0, k),
                                    b_block_buf,
                                    b_thread_desc_,
-                                   make_tuple(I0, I0, I0, I0),
+                                   make_tuple(n0, I0, I0, I0),
                                    b_thread_buf);
-
-                static_for<0, KPerBlock, KPack * xdlops_gemm.K0PerXdlops>{}([&](auto k) {
-                    vector_type<FloatAB, KPack> a_thread_vec;
-                    vector_type<FloatAB, KPack> b_thread_vec;
-
-                    static_for<0, KPack, 1>{}([&](auto i) {
-                        a_thread_vec.template AsType<FloatAB>()(i) = a_thread_buf
-                            [Number<a_thread_desc_.CalculateOffset(make_tuple(0, 0, 0, k + i))>{}];
-                        b_thread_vec.template AsType<FloatAB>()(i) = b_thread_buf
-                            [Number<b_thread_desc_.CalculateOffset(make_tuple(0, 0, 0, k + i))>{}];
+            });
+            __builtin_amdgcn_sched_barrier();
+            // NOTE: sync thread at the start of each MAC cluster except for the first MAC cluster
+            // A) we want waves in a workgroup in sync to prevent waves from other workgroups hijacking MAC resource
+            // B) the barrier after the last local read also acts as safeguard against data hazard
+            if constexpr ( int(k) != 0 || KPerInnerLoop == KPerBlock )
+            {
+                asm volatile("s_barrier" ::);
```

Comment:
There is lgkmcnt(0) here, is it always correct?

Reply:
The compiler will add lgkmcnt as appropriate, for example:

```
	ds_read_b128 v[18:21], v6 offset:4128
	ds_read_b128 v[22:25], v6 offset:5152
	ds_read_b128 v[26:29], v5 offset:8240
	ds_read_b128 v[30:33], v5 offset:8752
	ds_read_b128 v[34:37], v5 offset:10320
	ds_read_b128 v[38:41], v5 offset:10832
	; sched barrier
	;;#ASMSTART
	s_barrier
	;;#ASMEND
	; sched barrier
	s_waitcnt lgkmcnt(3)
	v_mfma_i32_32x32x8i8 a[32:47], v10, v26, a[32:47]
	v_mfma_i32_32x32x8i8 a[32:47], v11, v27, a[32:47]
	v_mfma_i32_32x32x8i8 a[32:47], v12, v28, a[32:47]
	v_mfma_i32_32x32x8i8 a[32:47], v13, v29, a[32:47]
```

Reply:
Per discussion, it is a practically safe but theoretically unsafe approach. It's been fixed in newer commits.